### ISSUE TITLE
Trigger of the Snippets

### ISSUE THREAD

**Issue by clsmt** [area/language-latex#119]
I have been using this package for a while now. And I am bothered by this issue more and more. 

I would **strongly** recommend using \begin as trigger as opposed to just "begin". I would also strongly recommend revising the trigger of "legal" and "lorem" to something like "\legal" or "\lorem". 

We are very used to type \begin{equation} to start an equation. It makes sense that now we just have to type \be... and \begin pops up as an option. Now an "enter" produce the full equation commands. But as of now, when I type \begin, \begin{equation} pops up, and an "enter" would produce something like "\\begin". I have to go back and erase the first "\".
Sure, if you are used to just "begin" gives you "\begin", it saves you a character, but experienced latex writers would have to change their habbit. If god forbidden they later on have to use other editors (for example writing something online) for latex, they have to change their habbit again. What a pain.

As for the "legal" or "lorem" trigger. The problem: anytime I write a word starting with "l", these two options pops up, and swiftly disappears when I continue to write whatever word I was writing. This is kind of cool in the first 5 seconds. After a while, they became very, very distracting. I believe most of experienced latex writer would agree with me on this one.

When I am writting latex, I want to be absolutely focused. I don't want to change any habits unnecessarily. I don't want any funky menus pop up and disappear all the time. I'd love it if useful snippets show up when I needed them; otherwise, I want the entire environment to be quite and still.

**Comment by gnat79** [comment]
I agree completely - I spend a lot of energy deciding whether to type "begin" or "\begin" depending on whether I want matching or not. I could save a lot of time by always typing "\begin".

**Comment by Aerijo** [comment]
Not to mention that latex is all about typing English. The word "begin" (and other snippet starters) could come up a lot, and it is annoying to see the snippet suggestion every time. Restricting it to when the backslash character is used will prevent this, and be more consistent with author's habits.

**Comment by Aerijo** [comment]
@clsmt I now realise why they are this way: the autocomplete-snippets package does not support punctuation in snippet prefixs. If you want a workaround for this issue, do the following

- Run the command `git clone https://github.com/atom/autocomplete-snippets`. 
- Make a link to this folder in the `.atom/packages` folder (one is probably automatically made in `.atom/dev/packages`. This one will only activate in `dev` mode). This step will allow your local copy of the package to take priority.
- Now, on your local copy, navigate to `autocomplete-snippets/lib/snippets-provider.coffee`.
- Make the changes shown in the [commit link here](https://github.com/ezuhaib/autocomplete-snippets/commit/00d31fd9df735fa16c67954f0f93b929bc10c22d)
- In addition to this, change line 27 to `regex = /[.]*[\w0-9_\-//]+$/`. This allows `\` to start a snippet as well, which is what we want.
- Reload Atom, and the changes should take effect. Now you can define your own snippets that begin with `\`.

I will try and get this behaviour working by default, but I don't know much about *hacking Atom* yet. The developers of the `autocomplete-snippets` package only want commits that work perfectly, and I guess our suffering isn't good enough for them ¯\\\_(ツ)\_/¯

**Comment by yudai-nkt** [comment]
I **DO** agree with including backslashes to snippet trigger. However, as can be seen in atom/autocomplete-snippets#67, snippets starting with non-letter are not autocompleted (snippet itself should work). That's why we can't do that.

**Comment by Aerijo** [comment]
This feature would be nice, but would have to be done by the `autocomplete-snippets` package, so I'm adding a `wontfix` label to this issue.

**Comment by Goysa2** [comment]
@Aerijo Hi! The autocomplete-snippet package has been update and the problem cannot be solved using the solution you provided. Any advice on how to fix it now?

**Comment by Aerijo** [comment]
@Goysa2 Hey, are you perhaps referring to this change? https://github.com/atom/autocomplete-snippets/pull/92

If so, that is unrelated to the current issue. If not, can you point me to the change, because I cannot reproduce it using the latest version of Atom.

For the solution I provided, it seems I had a typo in it. Additionally, they have now converted to JavaScript (form CoffeeScript). A better set of steps would be:

- Run the command `git clone https://github.com/atom/autocomplete-snippets` in the command line
- Make a link to this folder in the `.atom/packages` folder. Do this by navigating to the downloaded package and running the command `apm link`
- Now, in the downloaded package, navigate to `autocomplete-snippets/lib/snippets-provider.js`.
- Delete the `getSuggestions` method and replace it with the following
```js
  getSuggestions({scopeDescriptor, editor, bufferPosition}) {
    let prefix = this.getPrefix(editor, bufferPosition)
    if (!(prefix != null ? prefix.length : undefined)) { return }
    const scopeSnippets = this.snippetsSource.snippetsForScopes(scopeDescriptor)
    return this.findSuggestionsForPrefix(scopeSnippets, prefix)
  }

  getPrefix(editor, bufferPosition) {
    let regex = /[\w0-9\-\_\\]+$/
    let line = editor.getTextInRange([[bufferPosition.row, 0], bufferPosition])
    let match = line.match(regex)
    return match ? match[0] : ''
  }
```
- Reload Atom. Now you can define your own snippets that begin with `\`.
